src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -128,12 +156,90 @@ func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.Unsign
 		outTree,
 		addr.BurntFundsActorAddr,
 		minerAddr, // TODO: may not exist
-		gasToFIL(sendRet.GasUsed(), message.GasPrice()),
+		_gasToFIL(sendRet.GasUsed(), message.GasPrice()),
 	)
 
 	return outTree, sendRet
 }
 
-func gasToFIL(gas msg.GasAmount, price msg.GasPrice) actor.TokenAmount {
+func _applyError(errCode exitcode.SystemErrorCode) msg.MessageReceipt {
+	// TODO: should this gasUsed value be zero?
+	// If nonzero, there is not guaranteed to be a nonzero gas balance from which to deduct it.
+	gasUsed := gascost.ApplyMessageFail
+	TODO()
+	return msg.MessageReceipt_MakeSystemError(errCode, gasUsed)
+}
+
+func _withTransferFundsAssert(tree st.StateTree, from addr.Address, to addr.Address, amount actor.TokenAmount) st.StateTree {
+	// TODO: assert amount nonnegative
+	retTree, err := tree.Impl().WithFundsTransfer(from, to, amount)
+	if err != nil {
+		panic("Interpreter error: insufficient funds (or transfer error) despite checks")
+	} else {
+		return retTree
+	}
+}
+
+func _gasToFIL(gas msg.GasAmount, price msg.GasPrice) actor.TokenAmount {
 	return actor.TokenAmount(util.UVarint(gas) * util.UVarint(price))
 }
+
+// Builds a message for paying block reward from the treasury account to a miner owner.
+func _makeBlockRewardMessage(state st.StateTree, minerActorAddr addr.Address) msg.UnsignedMessage {
+	// networkTreasuryActor := loadActor(NetworkTreasuryActorAddress)
+	// minerActor := loadActorSubstate(minerActorAddr)
+	// minerWorker := loadActor(minerActor.Info.Worker)
+	//return &UnsignedMessage_I{
+	//	From_:       minerActor.Info.Worker,
```

Comment:
In general, I'm not sure we want to have the treasury send money in response to messages sent to it by other actors. Could this instead be the execution of a message whose sender is the treasury, and whose accompanying token value is the block reward?

Reply:
Treasury actor still needs to know in some way who mines the block tho, how will this be done?

Reply:
Top-level message sender must be an account actor, which are the only actors with meaningful CallSeqNum etc, unless we want to break that invariant (I'm not sure what depends on it). Yes, when we write it it will need some protection against being invoked manually. 

But in practice, miners are not going to reach consensus with another that's stealing extra block reward. 

We could invent a new account actor that's solely for the purpose of being the sender of these messages (but it can't have code, which we're now adding to the treasury actor).

Reply:
In either case, can I do it in follow-up?